File: ios/StationToStation/Data/Gossip/GossipWire.swift
import Foundation

// The wire layer of the gossip channel (#417), and the cross-platform contract two phones
// have to agree on byte for byte before a single check-in crosses between them. Pure Swift
// with no CoreBluetooth in sight, exactly as `CardWire.swift` is pure against
// `BleExchange.swift` (ADR-0001) — and for the harder reason as well: the transport above
// it cannot be asserted without two phones and a gig, so everything that *can* be decided
// from bytes alone is decided here, where XCTest reaches it.
//
// Android's twin is `data/GossipWire.kt`, and the assertions in `GossipWireTests` are the
// assertions Android runs on the JVM.
//
// Wire shape:
//   advertisement  — the 128-bit service UUID, and nothing else. See below.
//   token (read)   — this device's `gossipTokenOffer`, one token per line.
//   inbox (write)  — the writer's own token for the resolved pair, then its batch.
//   outbox (read)  — the reader's token as the peripheral computed it, then its batch.
//
// **Why the token is not in the advertisement, on this platform.** It would be the obvious
// place, and on Android it is one. CoreBluetooth's `startAdvertising` honours exactly two
// keys — a local name and a list of service UUIDs — so an iPhone cannot put arbitrary bytes
// in an advertisement at all (`BleExchange.swift` records the same fact for the Exchange's
// display name). Worse, a *backgrounded* iPhone drops the local name entirely and moves its
// service UUIDs into a special "overflow" area that only another iOS device explicitly
// scanning for that exact UUID can see. So on iOS the token cannot ride the advertisement,
// and it rides the first GATT read instead. The derivation is identical to Android's — that
// is the part that must match — and only the carrier differs, the same way the Exchange's
// name rides manufacturer data on one platform and a local name on the other.
//
// **Framing, since a batch does not fit in an MTU.** Both directions are a byte stream over
// one attribute, and the two directions are framed differently because GATT frames them
// differently:
//
//   * A **read** is a long read: the reader asks again at a rising offset until it gets a
//     short answer. The whole payload is addressable, so the writer serves slices of it and
//     the offset means what it says (`sliceForOffset`).
//   * A **write** is a sequence of appends, each at offset 0, terminated by a zero-length
//     write. Not offsets: CoreBluetooth does not perform prepared writes and silently
//     truncates a `writeValue` past the MTU, so the sender chunks by hand and every chunk
//     arrives at the peripheral looking like the start of the value. Append-and-terminate is
//     the one framing that reads identically on both platforms. Android's twin must chunk its
//     writes the same way — sequential, offset 0, empty write to finish — or an iPhone will
//     see only the last chunk of every batch it is handed.
//
// A zero-length write is therefore meaningful and never ignored: it is what says "that was
// the whole batch, you may prepare the reply now".

/// Fixed on both platforms, and deliberately not the Exchange's service. Change either and
/// the phones stop seeing each other.
///
/// A separate service, not another characteristic on `exchangeServiceUUIDString`, because the
/// two channels have opposite lifetimes: the Exchange advertises only while a human is on its
/// screen (ADR-0016) and this one advertises in the background (ADR-0019). One service would
/// mean a background scanner waking for every Exchange in the room, and an Exchange scanner
/// listing every pocket in the venue as a peer to tap.
let gossipServiceUUIDString = "7b7e6f2a-7601-4b1a-9e2c-2a6f6f0b7721"

/// Read first, by the connecting side: whose Contacts this device could be.
let gossipTokenCharacteristicUUIDString = "7b7e6f2a-7601-4b1a-9e2c-2a6f6f0b7722"

/// Written second, by the connecting side, once it knows who it is talking to: its token for
/// this pair, then everything it has to offer.
let gossipInboxCharacteristicUUIDString = "7b7e6f2a-7601-4b1a-9e2c-2a6f6f0b7723"

/// Read third: what the advertising side has to offer back, prepared once the write above
/// told it who is asking. Empty for a connection that never resolved to a **Contact** — a
/// stranger gets a zero-length read, not a batch and not an error.
let gossipOutboxCharacteristicUUIDString = "7b7e6f2a-7601-4b1a-9e2c-2a6f6f0b7724"

/// The most bytes read or accepted in one direction of one meeting.
///
/// `gossipMaxBatch` already bounds how many messages are *judged*; this bounds what is
/// allocated before anything has been judged at all, which is the part a peer controls. A
/// full 64-message batch is around 22 kB, so this is roughly one and a half of those — over
/// BLE, generous.
let gossipMaxWireBytes = 32 * 1024

/// The header every token offer starts with.
let gossipTokenListV1 = "station-to-station/gossip-tokens/1"

/// The header every batch starts with. A version in the string so a later envelope is a
/// different document rather than an ambiguous one, exactly as `gossipPayloadV1` is.
let gossipBatchV1 = "station-to-station/gossip-batch/1"

/// One peer's whole offer: who it says it is on this edge, and what it is handing over.
///
/// `token` is not trusted as an identity by this file — `gossipResolveToken` turns it into a
/// **Contact** key, and `gossipStormGate` decides what that key is worth. This type only
/// says the bytes parsed.
struct GossipBatch: Equatable {
    var token: String
    var messages: [GossipCheckIn]
}

/// The token offer, as it goes over the read.
///
/// Line-separated because a token is 32 hex characters and can contain nothing else, so
/// there is no escaping problem to get wrong on one platform and not the other.
func encodeGossipTokens(_ tokens: [String]) -> Data {
    Data(([gossipTokenListV1] + tokens.filter(isSafeGossipToken)).joined(separator: "\n").utf8)
}

/// Nil for anything that is not a token offer. Malformed lines are dropped rather than
/// failing the whole read: a peer running a later build may publish something this one does
/// not understand, and the tokens it *does* understand are still worth resolving.
func decodeGossipTokens(_ data: Data) -> [String]? {
    guard data.count <= gossipMaxWireBytes else { return nil }
    var lines = String(decoding: data, as: UTF8.self).split(separator: "\n",
                                                            omittingEmptySubsequences: false)
    guard lines.first == gossipTokenListV1[...] else { return nil }
    lines.removeFirst()
    return lines.map(String.init).filter(isSafeGossipToken)
}

/// A batch as it goes over the wire: the header, the sender's token for this edge, then one
/// tab-separated line per message.
///
/// Tab-separated, and safe to be: a `messageId` is hex, a `gigId` has been through
/// `isSafeGossipId`, a key and a signature are base64, and the times are decimals — none of
/// which can contain a tab or a newline. The fields that could (`gigId`, `checkedInBy`) are
/// checked anyway, here and in `gossipPayload`, because "cannot happen" is how a separator
/// injection gets written.
///
/// Times are epoch **seconds**, decimal — the same representation `gossipPayload` signs over,
/// so a message that survives this encoding round-trips to the identical id and the identical
/// signature check. Nil for a message that cannot be canonically encoded at all, which is the
/// same message `gossipPayload` refuses.
func encodeGossipBatch(token: String, messages: [GossipCheckIn]) -> Data? {
    guard isSafeGossipToken(token) else { return nil }
    var lines = [gossipBatchV1, token]
    for message in messages.prefix(gossipMaxBatch) {
        guard let line = gossipBatchLine(message) else { return nil }
        lines.append(line)
    }
    return Data(lines.joined(separator: "\n").utf8)
}

private func gossipBatchLine(_ message: GossipCheckIn) -> String? {
    guard isSafeGossipId(message.gigId), !message.messageId.isEmpty,
          !message.checkedInBy.isEmpty, !message.signature.isEmpty,
          let checkedInAt = gossipEpochSeconds(message.checkedInAt),
          let expiresAt = gossipEpochSeconds(message.expiresAt)
    else { return nil }
    let fields = [message.messageId, message.gigId, message.checkedInBy,
                  String(checkedInAt), String(expiresAt), message.signature]
    guard fields.allSatisfy({ !$0.contains("\t") && !$0.contains("\n") }) else { return nil }
    return fields.joined(separator: "\t")
}

/// Nil for anything that is not a batch. A malformed *line* is dropped, not fatal, for the
/// same forward-compatibility reason `decodeGossipTokens` drops one — and because a peer that
/// can invalidate a whole handover with one bad line is a peer that can stop a night's
/// check-ins reaching anybody.
///
/// Nothing here judges a message. Every field it produces is still the author's claim, and
/// `gossipStormGate` is the only thing that decides otherwise: this refuses bytes it cannot
/// read, and passes on bytes it can.
func decodeGossipBatch(_ data: Data) -> GossipBatch? {
    guard data.count <= gossipMaxWireBytes else { return nil }
    var lines = String(decoding: data, as: UTF8.self).split(separator: "\n",
                                                            omittingEmptySubsequences: false)
    guard lines.count >= 2, lines[0] == gossipBatchV1[...] else { return nil }
    let token = String(lines[1])
    guard isSafeGossipToken(token) else { return nil }
    lines.removeFirst(2)
    // Truncated at `gossipMaxBatch` rather than rejected wholesale: the gate rejects the
    // overflow as `.batchLimit` anyway, and refusing the read would throw away the 64
    // messages that were fine along with the ones over the line.
    let messages = lines.prefix(gossipMaxBatch).compactMap(gossipMessageFromLine)
    return GossipBatch(token: token, messages: messages)
}

private func gossipMessageFromLine(_ line: Substring) -> GossipCheckIn? {
    let fields = line.split(separator: "\t", omittingEmptySubsequences: false).map(String.init)
    guard fields.count == 6,
          let checkedInAt = Int64(fields[3]), let expiresAt = Int64(fields[4]),
          abs(checkedInAt) <= gossipMaxEpochSecond, abs(expiresAt) <= gossipMaxEpochSecond,
          !fields[0].isEmpty, !fields[2].isEmpty, !fields[5].isEmpty
    else { return nil }
    return GossipCheckIn(
        messageId: fields[0],
        gigId: fields[1],
        checkedInBy: fields[2],
        checkedInAt: Date(timeIntervalSince1970: TimeInterval(checkedInAt)),
        expiresAt: Date(timeIntervalSince1970: TimeInterval(expiresAt)),
        signature: fields[5]
    )
}

/// My own arrival, ready to travel: the message a check-in mints for the gossip channel.
///
/// The one place a message is *authored* rather than relayed, so it is the one place the id
/// and the signature are produced rather than checked. Built through `gossipPayload` and
/// `gossipMessageId` — the same functions `gossipStormGate` will recompute on the far end,
/// which is what guarantees a message this device mints is a message the other twin accepts.
///
/// `expiresAt` is the gig's own night end where this device knows the date (`gossipExpiry`),
/// and `gossipMaxLifetime` from the check-in where it does not. Never longer: the ceiling in
/// `gossipStormGate` would cap it on arrival anyway, and a claim nobody honours is a claim
/// not worth signing.
///
/// `sign` is `ContactIdentity.sign` in the app and a test's own key in a test — the same
/// injection `gossipStormGate` takes for `verify`, for the same reason: the Secure Enclave
/// does not exist off a phone, and the envelope's shape must be assertable anyway.
func gossipCheckInMessage(gigId: String, gigDate: String?, publicKey: String, now: Date,
                          calendar: Calendar = .current,
                          sign: (Data) -> Data?) -> GossipCheckIn? {
    guard isSafeGossipId(gigId), !publicKey.isEmpty, let seconds = gossipEpochSeconds(now)
    else { return nil }
    let checkedInAt = Date(timeIntervalSince1970: TimeInterval(seconds))
    let night = gigDate.flatMap { gossipExpiry(gigDate: $0, calendar: calendar) }
    let expiresAt = night ?? checkedInAt.addingTimeInterval(gossipMaxLifetime)
    guard expiresAt > checkedInAt else { return nil }
    var message = GossipCheckIn(messageId: "", gigId: gigId, checkedInBy: publicKey,
                                checkedInAt: checkedInAt, expiresAt: expiresAt, signature: "")
    guard let payload = gossipPayload(message), let signature = sign(payload),
          let messageId = gossipMessageId(message)
    else { return nil }
    message.messageId = messageId
    message.signature = signature.base64EncodedString()
    return message
}

File: android/app/src/main/java/io/github/magnusencoded/stationtostation/data/gossip/GossipWire.kt
package io.github.magnusencoded.stationtostation.data.gossip

import io.github.magnusencoded.stationtostation.data.GOSSIP_MAX_EPOCH_SECOND
import io.github.magnusencoded.stationtostation.data.GossipCheckIn
import java.time.Instant
import java.util.Base64

/**
 * What one device hands another over the gossip channel, as bytes (#416).
 *
 * Pure, and deliberately hand-built rather than left to a serialization library, for the
 * reason [gossipPayload][io.github.magnusencoded.stationtostation.data.gossipPayload]
 * gives: iOS (#417) has to produce the same bytes, and "whatever the two platforms' JSON
 * encoders happen to agree on" is not a specification. Tab-separated fields, newline-
 * separated records — a base64 value contains neither, and a gig id has been through
 * [isSafeGossipId][io.github.magnusencoded.stationtostation.data.isSafeGossipId], so no
 * field can carry its own separator.
 *
 * **The transport is push-only, and this is the whole of it.** Nothing is read back from a
 * peer: every device runs both halves of the radio, so a device with something to say
 * connects and writes, and a device with nothing to say never has to be trusted for
 * anything. That removes the authenticated-read direction entirely — there is no "prove
 * yourself to me so I can believe what you hand back", only "prove yourself to me before I
 * read what you pushed".
 */

/** The header a **Pass** starts with. Version in the string, same reasoning as elsewhere. */
const val GOSSIP_PASS_V1 = "station-to-station/gossip-pass/1"

/**
 * The domain separator for the possession proof.
 *
 * The **Contact** identity key also answers the LAN reconcile challenge (#265) and signs
 * gossip payloads
 * ([GOSSIP_PAYLOAD_V1][io.github.magnusencoded.stationtostation.data.GOSSIP_PAYLOAD_V1]).
 * Three uses, three prefixes, so a signature made for one can never be presented as an
 * answer to another. Reconcile's nonce is a certificate fingerprint — raw digest bytes —
 * so producing one that begins with this ASCII prefix would take a preimage, not a choice.
 */
const val GOSSIP_AUTH_V1 = "station-to-station/gossip-auth/1"

/**
 * How many bytes of nonce a listener demands back.
 *
 * Thirty-two, matching the digest the signature is taken over anyway. The nonce exists so
 * that a recording of yesterday's **Pass** cannot be replayed as today's identity; the only
 * property that has to hold is that a listener never issues the same one twice, which at
 * this width it will not.
 */
const val GOSSIP_NONCE_BYTES = 32

/**
 * The hard ceiling on one **Pass**, in bytes.
 *
 * Not the same bound as
 * [GOSSIP_MAX_BATCH][io.github.magnusencoded.stationtostation.data.GOSSIP_MAX_BATCH] and
 * not a substitute for it: that one is the gate's rule about how many messages are *read*,
 * this one is the transport's rule about how much memory a peer's write is allowed to cost
 * before anything has been decided at all. A hostile **Contact** must not be able to make
 * this device hold a megabyte because it opened a GATT connection.
 *
 * Sixty-four messages at their realistic worst — a 64-character id, a 64-character gig id,
 * a ~124-character key, two timestamps and a ~96-character signature — is a little over
 * 32 KB, so this leaves room without leaving a hole. Anything larger is refused whole
 * rather than truncated: half a **Pass** is not a **Pass**.
 */
const val GOSSIP_MAX_WIRE_BYTES = 40_000

/** One **Pass**: who claims to be pushing, their proof, and what they push. */
data class GossipPass(
    /** The pushing peer's identity key, base64 X.509 SubjectPublicKeyInfo. */
    val from: String,
    /** Base64 DER over [gossipAuthPayload] of the nonce this device issued. */
    val proof: String,
    val batch: List<GossipCheckIn>,
)

/**
 * The bytes a peer signs to prove it holds the key it claims.
 *
 * The nonce is base64'd rather than concatenated raw so that the payload is text throughout
 * and the two platforms cannot disagree about byte order or padding in the middle of a
 * signed value.
 */
fun gossipAuthPayload(nonce: ByteArray): ByteArray =
    "$GOSSIP_AUTH_V1\n${Base64.getEncoder().encodeToString(nonce)}".toByteArray(Charsets.UTF_8)

/**
 * Null for anything that cannot be encoded unambiguously — a key or signature carrying a
 * separator, or a message [gossipPayload][io.github.magnusencoded.stationtostation.data.gossipPayload]
 * itself would refuse. A message that cannot be written is dropped from the batch rather
 * than failing the whole **Pass**: the rest of the night's news is still worth pushing.
 */
fun encodeGossipPass(pass: GossipPass): ByteArray? {
    if (pass.from.isBlank() || pass.proof.isBlank()) return null
    if (!isWireSafe(pass.from) || !isWireSafe(pass.proof)) return null
    val records = pass.batch.mapNotNull(::encodeRecord)
    val text = (listOf(GOSSIP_PASS_V1, "${pass.from}\t${pass.proof}") + records)
        .joinToString("\n")
    val bytes = text.toByteArray(Charsets.UTF_8)
    return if (bytes.size > GOSSIP_MAX_WIRE_BYTES) null else bytes
}

/**
 * Null for anything that is not a **Pass**: the wrong header, no claim line, an oversized
 * write, or bytes that are not UTF-8.
 *
 * Individual records that do not parse are **skipped, not fatal**. The alternative — one
 * unreadable record discarding a peer's whole batch — hands any device in the chain a way
 * to stop a message it does not like by corrupting the one next to it. Nothing is trusted
 * either way: what survives here still has to get past the storm-gate, which recomputes
 * every id and checks every signature.
 */
fun decodeGossipPass(bytes: ByteArray?): GossipPass? {
    if (bytes == null || bytes.isEmpty() || bytes.size > GOSSIP_MAX_WIRE_BYTES) return null
    val lines = String(bytes, Charsets.UTF_8).split('\n')
    if (lines.size < 2 || lines[0] != GOSSIP_PASS_V1) return null
    val claim = lines[1].split('\t')
    if (claim.size != 2) return null
    val from = claim[0]
    val proof = claim[1]
    if (from.isBlank() || proof.isBlank()) return null
    return GossipPass(from, proof, lines.drop(2).mapNotNull(::decodeRecord))
}

private fun encodeRecord(message: GossipCheckIn): String? {
    val fields = listOf(
        message.messageId,
        message.gigId,
        message.checkedInBy,
        message.checkedInAt.epochSecond.toString(),
        message.expiresAt.epochSecond.toString(),
        message.signature,
    )
    if (fields.any { it.isEmpty() || !isWireSafe(it) }) return null
    return fields.joinToString("\t")
}

/**
 * One record back into a message, or null.
 *
 * The times are parsed with the same bound
 * [GOSSIP_MAX_EPOCH_SECOND][io.github.magnusencoded.stationtostation.data.GOSSIP_MAX_EPOCH_SECOND]
 * states, and for the same reason: a peer's decimal is a peer's decimal, and
 * [Instant.ofEpochSecond] throws on values the gate would otherwise never get to see.
 */
private fun decodeRecord(line: String): GossipCheckIn? {
    val f = line.split('\t')
    if (f.size != 6) return null
    if (f.any { it.isEmpty() }) return null
    val checkedInAt = f[3].toLongOrNull()?.let(::boundedInstant) ?: return null
    val expiresAt = f[4].toLongOrNull()?.let(::boundedInstant) ?: return null
    return GossipCheckIn(
        messageId = f[0],
        gigId = f[1],
        checkedInBy = f[2],
        checkedInAt = checkedInAt,
        expiresAt = expiresAt,
        signature = f[5],
    )
}

internal fun boundedInstant(epochSecond: Long): Instant? =
    if (epochSecond > GOSSIP_MAX_EPOCH_SECOND || epochSecond < -GOSSIP_MAX_EPOCH_SECOND) null
    else Instant.ofEpochSecond(epochSecond)

private fun isWireSafe(value: String): Boolean = !value.contains('\n') && !value.contains('\t')
